examples/tunables_limit_memory.rs

```diff
@@ -0,0 +1,131 @@
+use std::sync::Arc;
+use wasmer::{
+    imports, wat2wasm, Instance, Memory, MemoryError, MemoryType, Module, Pages, Store, TableType,
+    Tunables as BaseTunables,
+};
+use wasmer_compiler::Target;
+use wasmer_compiler_cranelift::Cranelift;
+use wasmer_engine::Tunables;
+use wasmer_engine_jit::JIT;
+use wasmer_vm::{Memory as MemoryTrait, MemoryStyle, Table, TableStyle};
+
+/// A custom tunables that allows you to set a memory limit
+struct LimitingTunables {
+    /// The maxium a linear memory is allowed to be (in Wasm pages, 65 KiB each).
+    /// Since Wasmer ensures there is only none or one memory, this is practically
+    /// an upper limit for the guest memory.
+    max_memory: Pages,
+    /// The base implementation we delegate all the logic to
+    base: BaseTunables,
+}
+
+impl LimitingTunables {
+    pub fn for_target(target: &Target, limit: Pages) -> Self {
+        Self {
+            max_memory: limit,
+            base: BaseTunables::for_target(target),
+        }
+    }
+}
+
+impl Tunables for LimitingTunables {
+    /// Construct a `MemoryStyle` for the provided `MemoryType`
+    ///
+    /// Delegated to base.
+    fn memory_style(&self, memory: &MemoryType) -> MemoryStyle {
+        self.base.memory_style(memory)
+    }
+
+    /// Construct a `TableStyle` for the provided `TableType`
+    ///
+    /// Delegated to base.
+    fn table_style(&self, table: &TableType) -> TableStyle {
+        self.base.table_style(table)
+    }
+
+    /// Create a memory given a memory type
+    ///
+    /// The requested memory type is validated, adjusted to the limited and then passed to base.
+    fn create_memory(
+        &self,
+        requested: &MemoryType,
+        style: &MemoryStyle,
+    ) -> Result<Arc<dyn MemoryTrait>, MemoryError> {
+        if requested.minimum > self.max_memory {
+            return Err(MemoryError::Generic(
+                "Minimum of requested memory exceeds the allowed memory limit".to_string(),
+            ));
+        }
+
+        if let Some(max) = requested.maximum {
+            if max > self.max_memory {
+                return Err(MemoryError::Generic(
+                    "Maximum of requested memory exceeds the allowed memory limit".to_string(),
+                ));
+            }
+        }
```

Comment:
One thing, if we expect users to be able to handle these errors programmatically, we should use semantic error values rather than a String via `MemoryError::Generic` if this error is something fatal then a String is fine

Reply:
Oh, I just realized that this is an example 😆 , this is neat!

The same still holds though, we may want to add error values for these if we want users to be able to handle them better.  We can adjust this this after shipping this example though, so let's ship it as-is for now and then follow up with possibly adding error values that are easier to match on.

Reply:
I'm happy wither way. I could create the error types, but then again the whole concept of memory limits is a customization and the errors also belong in the customization. So maybe the question is whether or not the error type must be customizable in `Result<Arc<dyn MemoryTrait>, MemoryError>`.